linode/obj/resource.go

```diff
@@ -120,16 +154,31 @@ func diffResource(
 // specified bucket via the *schema.ResourceData, then it calls
 // readResource.
 func putObject(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := helper.S3ConnFromResourceData(d)
+	// put happens before read which sets the endpoint, compute to be safe
+	endpoint := d.Get("endpoint").(string)
+	if endpoint == "" {
+		var err error
+		if endpoint, err = helper.ComputeS3Endpoint(ctx, d, meta); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	bucket := d.Get("bucket").(string)
+	accessKey := d.Get("access_key").(string)
+	secretKey := d.Get("secret_key").(string)
+	key := d.Get("key").(string)
+
+	client, err := helper.S3Connection(endpoint, accessKey, secretKey)
+	if err != nil {
+		return diag.FromErr(err)
+	}
```

Comment:
nit: This logic to compute endpoint seems being used in read, delete, and put. We may consider putting it into a helper function to reuse.

Reply:
@yec-akamai updated to go back to the connection form data concept, only tricky part is pulling the endpoint from the s3 client config instead of the local string